validator/spanner_location.rego

```diff
@@ -0,0 +1,63 @@
+#
+# Copyright 2019 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+package templates.gcp.GCPSpannerLocationConstraintV1
+
+import data.validator.gcp.lib as lib
+
+deny[{
+	"msg": message,
+	"details": metadata,
+}] {
+	constraint := input.constraint
+	lib.get_constraint_params(constraint, params)
+
+	asset := input.asset
+
+	# Applies to spanner instances only
+	asset.asset_type == "spanner.googleapis.com/Instance"
+
+	# Retrieve the list of allowed locations - default to empty set
+	locations := cast_set(lib.get_default(params, "locations", {}))
+
+	# The asset raises a violation if location_is_valid is evaluated to false
+not location_is_valid(	asset, locations)
+
+	message := sprintf("%v is in violation.", [asset.name])
+	metadata := {
+		"resource": asset.name,
+		"valid-locations": locations,
+	}
+}
+
+###########################
+# Rule Utilities
+###########################
+
+location_is_valid(asset, locations) {
+	# ensure we have a data object
+	resource := asset.resource.data
+
+	# Retrieve the location 
+	instance_config_parts := split(resource.config, "/")
+	resouce_location := instance_config_parts[3]
+
+	# iterate through the locations
+	location := locations[_]
+
+	# the resource location is valid if it matches one of the passed locations
+re_match(	location, resouce_location)
```

Comment:
This formatting looks off.

Reply:
same comment as earlier - make format did this :(

Reply:
Could you file an issue on the OPA repo so it's tracked upstream and link to it in a comment?

Reply:
There was one open for "not" already, I added a comment about re_match too: https://github.com/open-policy-agent/opa/issues/1560